lib/split/user.rb

```diff
@@ -10,20 +10,30 @@ def initialize(context, adapter=nil)
       @user = adapter || Split::Persistence.adapter.new(context)
     end
 
+    def [](key)
+      if Split::Experiment.key_without_version_for_experiment_key(key.to_s) == key.to_s
+        #if the key doesn't have version, return one with version
+        #this might not be needed anymore
+        @user[key_for_experiment_name(key)]
+      else
+        @user[key]
+      end
+    end
+
     def cleanup_old_experiments!
       keys_without_finished(user.keys).each do |key|
-        experiment = ExperimentCatalog.find key_without_version(key)
+        experiment = ExperimentCatalog.find Split::Experiment.key_without_version_for_experiment_key(key)
         if experiment.nil? || experiment.has_winner? || experiment.start_time.nil?
           user.delete key
-          user.delete Experiment.finished_key(key)
+          user.delete experiment.finished_key(key) if experiment
         end
       end
     end
 
     def max_experiments_reached?(experiment_key)
       if Split.configuration.allow_multiple_experiments == 'control'
         experiments = active_experiments
-        count_control = experiments.count {|k,v| k == experiment_key || v == 'control'}
+        count_control = experiments.count {|k,v| experiment_key =~ /#{k}(?:\:\d+)?/ || v == 'control'}
         experiments.size > count_control
```

Comment:
This whole function is kind of dense, but *if* I'm reading this correctly there might be a bug here. 

Here's what I this function is doing:
If `allow_multiple_experiments == 'control'` a user can be in many experiments, but they can only be in a single experiment where they are not in the 'control' Alternative. The previous check counted the total number of experiments where the user was in the 'control' Alternative plus one if the user was in the `experiment_key` experiment outside of the control group. I'm guessing there's an assumption here `experiment_key` is the experiment where you aren't in the control?  Is that a correct assumption to make though? I *suspect* what's really needed here is `experiments.size >= (count_control + 1)`, and then only count where `v == 'control'`.

Reply:
Sorry, I'm still trying to work my way through this.  I may have been totally off.

So we have `experiments.size > count_control`.  Basically, if we have more active experiments then we have experiments where we're the 'control' group then there is at least one experiment where we're not in a control so we shouldn't be non-control for anything else. 

But I get a bit lost of the meaning of the `experiment_key`. Because it's checking for this key inside of `experiments.count {|k,v| ....` the value of `count_control` is more than just "number of experiments where you're the control".

Is this a correct equivalent?

```ruby
experiments = active_experiments
non_controls = experiments.select {|k,v| v != 'control' }
if 0 == non_controls.length
  return false
elsif (1 == non_controls.length) && non_controls[0].has_key?(experiment_key)
  return false
else
  return true
end
```

Maybe that's not any better tbh :man_shrugging: 

I guess what I'm really confused about, is why you'd want to say: `max_experiments_reached?` is true if there's any non-control group is assigned, unless that assignment is for the specified experiment.

Reply:
At the very least the variable names are accurate in your version. I think the intention behind such a confusing calculation is to have the idea of `max_experiments?` be true based on the experiment key in consideration. This is probably the point where we ask for trouble. The rest of the codebase no doubt relies on this bizarre behavior when deciding to rebucket.